solr/core/src/java/org/apache/solr/update/SolrCoreState.java

```diff
@@ -86,7 +103,45 @@ public boolean decrefSolrCoreState(IndexWriterCloser closer) {
     }
     return close;
   }
-  
+
+  /**
+   * Pauses all update requests to this core and waits (indefinitely) for all in-flight
+   * update requests to finish
+   */
+  public void pauseUpdatesAndAwaitInflightRequests() {
+    if (pauseUpdateRequests.compareAndSet(false, true)) {
+      inflightUpdatesCounter.register();
+      inflightUpdatesCounter.arriveAndAwaitAdvance();
+    }
+  }
+
+  /**
+   * Unpauses update requests to this core
+   */
+  public void unpauseUpdates() {
+    this.pauseUpdateRequests.set(false);
+  }
+
+  /**
+   * Registers in-flight update requests to this core.
+   *
+   * @return true if request was registered, false if update requests are paused
+   */
+  public boolean registerInFlightUpdate() {
+    if (pauseUpdateRequests.get()) {
+      return false;
+    }
+    inflightUpdatesCounter.register();
```

Comment:
Do we need to check for termination?

Reply:
According to the javadocs, register() is a no-op if phaser has terminated so we don't need to worry about it.

Reply:
Should we return an error in this case? Does termination imply that the core is closed?

Reply:
Maybe `return inflightUpdatesCounter.register() > 0`? Or `>=`, I'm not sure.

Reply:
Thanks for bringing my attention to termination. The usage of the phaser was wrong before because it would get terminated the moment there were no in-flight requests. We actually never need to terminate the phaser at all. Now, we override onAdvance() and disable termination such that the phase keeps advancing and wraps around after reaching 65536.